pkg/operator/controller/controller.go

```diff
@@ -112,58 +109,42 @@ func (r *reconciler) reconcile(request reconcile.Request) (reconcile.Result, err
 			// resource that got deleted async).
 			logrus.Infof("clusteringress %q not found and reconciliation will be skipped", request)
 		} else {
-			// Try again later.
-			logrus.Errorf("failed to get clusteringress %q: %v", request, err)
-			result.RequeueAfter = 10 * time.Second
+			errs = append(errs, fmt.Errorf("failed to get clusteringress %q: %v", request, err))
 		}
 		ingress = nil
 	}
 
 	caSecret, err := r.ensureRouterCACertificateSecret()
 	if err != nil {
-		return reconcile.Result{}, fmt.Errorf("failed to ensure CA secret: %v", err)
+		errs = append(errs, fmt.Errorf("failed to ensure CA secret: %v", err))
 	}
 
-	// Collect errors as we go.
-	errs := []error{}
-
-	// Only reconcile the ingress itself if it exists.
 	if ingress != nil {
 		dnsConfig := &configv1.DNS{}
 		if err := r.Client.Get(context.TODO(), types.NamespacedName{Name: "cluster"}, dnsConfig); err != nil {
-			logrus.Errorf("failed to get dns 'cluster': %v", err)
+			errs = append(errs, fmt.Errorf("failed to get dns 'cluster': %v", err))
 			dnsConfig = nil
 		}
 		infraConfig := &configv1.Infrastructure{}
 		if err := r.Client.Get(context.TODO(), types.NamespacedName{Name: "cluster"}, infraConfig); err != nil {
-			logrus.Errorf("failed to get infrastructure 'cluster': %v", err)
+			errs = append(errs, fmt.Errorf("failed to get infrastructure 'cluster': %v", err))
 			infraConfig = nil
 		}
-		if dnsConfig == nil || infraConfig == nil {
-			// For now, if the cluster configs are unavailable, defer reconciliation
-			// because weaving conditionals everywhere to deal with various nil states
-			// is too complicated. It doesn't seem too risky to rely on the invariant
-			// of the cluster config being available.
-			result.RequeueAfter = 10 * time.Second
-		} else {
-			// Ensure we have all the necessary scaffolding on which to place router
-			// instances.
-			err = r.ensureRouterNamespace()
-			if err != nil {
-				errs = append(errs, err)
-				return result, utilerrors.NewAggregate(errs)
-			}
 
-			if ingress.DeletionTimestamp != nil {
-				// Handle deletion.
-				if err := r.ensureIngressDeleted(ingress, dnsConfig); err != nil {
-					errs = append(errs, fmt.Errorf("failed to ensure ingress deletion: %v", err))
-				}
-			} else {
-				// Handle everything else.
-				if err := r.ensureRouterForIngress(ingress, caSecret, infraConfig, dnsConfig, &result); err != nil {
-					errs = append(errs, fmt.Errorf("failed to ensure clusteringress: %v", err))
-				}
+		// Ensure we have all the necessary scaffolding on which to place router instances.
+		if err = r.ensureRouterNamespace(); err != nil {
+			errs = append(errs, fmt.Errorf("failed to ensure router namespace: %v", err))
+		}
```

Comment:
Neither `ensureIngressDeleted` nor `ensureRouterForIngress` can do any work if the namespace does not exist, so we can skip them if `ensureRouterNamespace` fails.

Reply:
`ensureRouterNamespace` not only creates router namesapce but also creates service account, cluster role and cluster rolebindings. If namespace creation succeeds but fails on other cases, we could definitely perform work during ingress deletion and creation. If namespace itself fails, both `ensureIngressDeleted` and `ensureRouterForIngress` will bailout as expected.